### PR TITLE
v622: Fix pair offset/size calculation.

### LINKED ISSUE #6840: TClass for pair sometimes have the wrong offset/size
Recent updated had made TClass::GetClass always returns an instance for std::pair even if there is no dictionary for the pair or it has not been instanced yet in the interpreter (step that can lead to autoparsing).
However, the offset and size calculations (guesses) in those cases is inadequate and sometimes leads to incorrect use (read/write) of the content of collection.  See for example: https://github.com/cms-sw/cmssw/issues/30359#issuecomment-728121661

### PR DESCRIPTION
**Pull request by pcanal**
Fix #6840. i.e Improve handling for std::pair without a dictionary.

### CODE REVIEW

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
Failures are unrelated (pre-exisiting).  @Axel-Naumann This is ready to merge.

File: core/meta/inc/TClass.h [hunk 1]
```diff
@@ -222,7 +223,13 @@ friend class ROOT::Internal::TCheckHashRecursiveRemoveConsistency;
    ClassConvStreamerFunc_t fConvStreamerFunc;   //Wrapper around this class custom conversion Streamer member function.
    Int_t               fSizeof;         //Sizeof the class.
 
-           Int_t      fCanSplit;          //!Indicates whether this class can be split or not.
+   // Bit field
+   Int_t fCanSplit : 3;          //!Indicates whether this class can be split or not. Values are -1, 0, 1, 2
+
+   //! Indicates whether this class represent a pair and was not created from a dictionary nor interpreter info but has
+   //! compiler compatible offset and size (and all the info is in the StreamerInfo per se)
```
Comment (Axel-Naumann) [inline core/meta/inc/TClass.h:230]:
```suggestion
   /// Indicates whether this class represents a pair and was not created from a dictionary nor interpreter info but has
   /// compiler compatible offset and size (and all the info is in the StreamerInfo per se)
```

File: core/meta/inc/TClass.h [hunk 2]
```diff
@@ -535,7 +543,8 @@ friend class ROOT::Internal::TCheckHashRecursiveRemoveConsistency;
    static void           RemoveClass(TClass *cl);
    static void           RemoveClassDeclId(TDictionary::DeclId_t id);
    static TClass        *GetClass(const char *name, Bool_t load = kTRUE, Bool_t silent = kFALSE);
-   static TClass        *GetClass(const std::type_info &typeinfo, Bool_t load = kTRUE, Bool_t silent = kFALSE);
+   static TClass        *GetClass(const char *name, Bool_t load, Bool_t silent, size_t hint_pair_offset, size_t hint_pair_size);
+   static TClass        *GetClass(const std::type_info &typeinfo, Bool_t load = kTRUE, Bool_t silent = kFALSE, size_t hint_pair_offset = 0, size_t hint_pair_size = 0);
```
Comment (Axel-Naumann) [inline core/meta/inc/TClass.h:547]: As this is a public method, would it make sense to swap the last two parameters for backward compatibility?

Or, to break the interface for existing callers and make sure they don't silently get wrong results, to take an `array<size_t, 2>` or a `struct PairInfo_t` as last parameter?

File: io/io/src/TStreamerInfo.cxx [hunk 3]
```diff
@@ -5687,11 +5711,22 @@ TVirtualStreamerInfo *TStreamerInfo::GenerateInfoForPair(const std::string &firs
    gErrorIgnoreLevel = kError;
    i->BuildCheck(nullptr, kFALSE); // Skipping the loading part (it would leads to infinite recursion on this very routine)
    gErrorIgnoreLevel = oldlevel;
+   // In the state emulated, BuildOld would recalculate the offset and undo the offset update.
+   // Note: we should consider adding a new state just for this (the hints indicates that we are mapping a compiled class but
+   // then we would have to investigate all use of the state with <= and >= condition to make sure they are still appropriate).
+   if (hint_pair_size) {
+      i->GetClass()->SetClassSize(hint_pair_size);
+      i->GetClass()->fIsSyntheticPair = kTRUE;
```
Comment (Axel-Naumann) [inline io/io/src/TStreamerInfo.cxx:5719]: Can you assert that `!hint_pair_offset == !hint_pair_size`? Else things will go belly up - and that's certainly a possibility given that an existing interface was amended.

**Review by Axel-Naumann — APPROVED** [review]

File: core/meta/inc/TClass.h (hunk 2 above)
Comment (pcanal) [reply core/meta/inc/TClass.h:547]: The new argument are already additive (require to specify the default arguments).

File: io/io/src/TStreamerInfo.cxx (hunk 3 above)
Comment (pcanal) [reply io/io/src/TStreamerInfo.cxx:5725]: Sure

File: core/meta/inc/TClass.h (hunk 1 above)
Comment (pcanal) [reply core/meta/inc/TClass.h]: Done.

File: io/io/src/TStreamerInfo.cxx [hunk 4]
```diff
@@ -5678,6 +5678,12 @@ TVirtualStreamerInfo *TStreamerInfo::GenerateInfoForPair(const std::string &firs
    // This TStreamerInfo is then used as if it was read from a file to generate
    // and emulated TClass.
 
+   if (hint_pair_offset && hint_pair_offset == hint_pair_size) {
```
Comment (Axel-Naumann) [inline io/io/src/TStreamerInfo.cxx:5681]: I really meant to please add an assert on `(!hint_pair_offset == !hint_pair_size)`: it's checking that either both or non are `==0`. Your current `if` doesn't check that.

**Pull request merged**